### DISCUSSION TITLE
[Request] Slightly but powerful enhanced PowerDelta

### DISCUSSION THREAD

**Discussion by muhkuhgh** [arendst/Tasmota#14969] [Ideas]
To catch slowly incrasing or decreasing power changes more frequently without spamming too much.

If I get it right, PowerDelta at the moment works like that:

PowerDelta 104 => Every 4 watt difference will be reported.

That means

If we start on 100
105 - report
120 - report
117
114
111
108 and so on..
no report other than on telemetry period...

I think it could be very helpful, to have the differences accumulated... and not barely checked on every reported Power Value.. Because in most of the cases I guess changes of my desired PowerDelta in like a certain amount of time are also verry interesting..

**Example:** 

PowerDelta 104

We start on 100.
106 - report. Keep that 106 in mind.
107 - +1 no report alone.
108 - +1=+2 no report alone.
109 - +1=+3 no report alone.
110 - +1=+4 no report alone but seen from the last report we've seen an increase of our desired PowerDelta so we send an Report.

This should be farily easy to implement.. We just need two more variables, one to store the latest reported value and one to sum up differences on every new power value.. then one simple if routine to check the the difference of our accumulated differences minus our last reported value and that absolute value against our desired powerdelta (maybe PowerDeltaAcc)

I have many use cases, where the generation/load slow increases in decreases.. but I don't want to send/store every single watt change.. but i neither want to wait telemetry period of time to get reported of above shown example :D

Like this:

After Power Report:
store lastReportedPower, sumPowerDelta = 0

on every new power value do:
sumPowerDelta = sumPowerDelta + recent Power - recentPowerbefore 

check if abs(sumPowerDelta) > desired PowerDelta,
if yes:
report. set lastReportedPower and reset sumPowerDelta = 0.

Power 100 reported on start. sumPowerDelta = 0
next measured value 102 -> sumPowerDelta =  0 + recent Power - recentpowerbefore = 102-100 = 2
next measured value 101 -> sumPowerDelta = 2 + 101 - 102 = 1
next measured value 102 -> sumPowerDelta = 1+ 102 - 101 = 2
next measured value 104 -> sumPowerDelta = 2 + 104 - 102 = 4 => Report.

**Comment by sfromis** [comment]
You are right that gradually changing energy consumption level will not trigger PowerDelta. I've not had a situation where I lacked this, but in that case, I could have used rules to get extra reporting when wanted.

In general, Tasmota drivers aim to use as little RAM as possible, and thus avoid keeping track of values back in time.

**Reply by barbudor** [reply]
I've seen situations where slowly changing values would not trigger PowerDelta when a "reasonnable" threshold was use.
The workaround was to use a much lower threshold levels.
A recent PR allows to trigger now on a change fo 1 W

Now I don't see why you would need to store both the last reported value and the sum of difference. This sounds redundant for me.
Either one or the other should be enough.

**Reply by sfromis** [reply]
A simplified extra trigger could be detecting n cycles with change in the same direction.

**Reply by stefanbode** [reply]
I also trapped into this. The user expectation in my case was that I get an additional report when the difference is reached compared to the last report. This silent shift is quite annoying. At the end you did not get information you expect. The current implementation is more or less an information if the speed of change trigger a limit. Quite unusual use-case

**Reply by sfromis** [reply]
Well, the current implementation is good for a very typical use case, things switching on and off. The use case of gradual change is IMO not as common.

**Comment by stefanbode** [comment]
Even if we change existing functionality in my mind we only need to change the compare value. Maybe one additional variable. No need to store this on flash. Memory is fine.

**Comment by TimKlebig** [comment]
May I please ask if there has been any update on the issue? I am facing the same problem and would appreciate any information you could share. Thank you very much.
